virtwho/config.py

```diff
@@ -685,29 +694,487 @@ def addConfig(self, config):
+    dictionary like. This object maintains a state attribute. The state shows if the configuration
+    section has passed validation, needs validation, or is invalid for some reason.
+    """
+    # The string representation of all default properties and values
+    DEFAULTS = ()
+    REQUIRED = ()
+
+    __marker = object()
+
+    def __init__(self, section_name, wrapper):
+        self.section_name = section_name
+        self._wrapper = wrapper
+        self.defaults = dict(type(self).DEFAULTS)
+        # Those properties that have yet to be validated
+        # Any item not in this set but in this ConfigSection is validated
+        self._unvalidated_keys = set()
+        self._invalid_keys = set()
+        # Holds all properties for this section
+        self._values = {}
+        # Add section defaults
+        for key, value in self.defaults.iteritems():
+            self._values[key] = value
+            self._unvalidated_keys.add(key)
+        self._update_state()
+
+    def iteritems(self):
+        for key in self:
+            yield (key, self._values[key])
+
+    def __iter__(self):
+        return iter(self._values.keys())
+
+    def _update_state(self):
+        if len(self._unvalidated_keys) > 0:
+            self.state = self._wrapper.State.NEEDS_VALIDATION
+        elif len(self._invalid_keys) > 0 or len(self._values) == 0:
+            self.state = self._wrapper.State.INVALID
+        else:
+            self.state = self._wrapper.State.VALID
+
+    def __delitem__(self, key):
+        if key in self:
+            if self.has_default(key):
+                return
+            del self._values[key]
+            if key in self._unvalidated_keys:
+                self._unvalidated_keys.remove(key)
+            if key in self._invalid_keys:
+                self._invalid_keys.remove(key)
+            self._update_state()
+        else:
+            raise KeyError('Unable to delete nonexistant property "%s"' % key)
+
+    def __setitem__(self, key, value):
+        if key not in self or (key in self and self[key] != value):
+            self._unvalidated_keys.add(key)
+            self._values[key] = value
+            self._update_state()
+
+    def __len__(self):
+        return len(self._values)
+
+    def __getitem__(self, key):
+        return self._values[key]
+
+    def validate(self, validation_messages=None):
+        validation_messages = validation_messages or []
+        # Do validation in subclasses
+
+        # Finally calls _update_state
+        self._update_state()
+        return validation_messages
+
+    def is_default(self, key):
+        return self.defaults[key] == self._values[key]
+
+    def has_default(self, key):
+        return key in self.defaults
+
+    def get(self, key, default=__marker):
+        try:
+            return self[key]
+        except KeyError as e:
+            if default is not self.__marker:
```

Comment:
To allow the default to be None I had to set the optional arg to something other than None.
This way if the user sets default to anything (so long as it is not self.__marker) and the given key is not in self, we'll return default.